File: App/Backend/src/services/invitation_service.py
import hashlib
import os
import secrets
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from Backend.src.core.security import hash_password
from Backend.src.repositories.teacher_repository import TeacherRepository
from Backend.src.repositories.user_repository import UserRepository
from Backend.src.services.auth_service import generate_uid
from Backend.src.services.email_service import (
    send_teacher_invite_email,
    send_teacher_welcome_email,
)

INVITE_EXPIRY_HOURS = int(
    os.getenv("TEACHER_INVITE_EXPIRY_HOURS", "48")
)


def _hash_token(token: str) -> str:
    return hashlib.sha256(token.encode()).hexdigest()


def create_teacher_invitation(
    db: Session,
    email: str,
    invited_by_uid: str,
    accept_url_base: str = ""
) -> dict:
    email = email.strip().lower()

    # Check active invitation
    existing = TeacherRepository.get_active_invitation_by_email(db, email)
    if existing:
        raise ValueError(
            "An active invitation already exists for this email. "
            "It will expire in less than 48 hours."
        )

    # Check user existence
    if UserRepository.get_by_email(db, email):
        raise ValueError("A user with this email already exists.")

    # Get admin/inviter name
    inviter = UserRepository.get_by_uid(db, invited_by_uid)
    inviter_name = inviter.username if inviter else "Admin"

    token = secrets.token_urlsafe(48)
    token_hash = _hash_token(token)
    expires_at = datetime.utcnow() + timedelta(hours=INVITE_EXPIRY_HOURS)

    invitation = TeacherRepository.create_invitation(
        db=db,
        email=email,
        token_hash=token_hash,
        expires_at=expires_at,
        invited_by_uid=invited_by_uid
    )

    email_sent = send_teacher_invite_email(
        to_email=email,
        invite_token=token,
        invited_by_name=inviter_name,
        accept_url_base=accept_url_base
    )

    if not email_sent:
        return {
            "message": (
                "Invitation was created, but the invitation email "
                "could not be sent."
            ),
            "invitation_id": invitation.invitation_id,
            "email": email,
            "expires_in_hours": INVITE_EXPIRY_HOURS
        }

    return {
        "message": f"Invitation sent to {email}.",
        "invitation_id": invitation.invitation_id,
        "email": email,
        "expires_in_hours": INVITE_EXPIRY_HOURS
    }


def accept_teacher_invitation(
    db: Session,
    token: str,
    username: str,
    password: str,
    name: str,
    phone_number: str | None = None,
    specialization: str | None = None,
    qualification: str | None = None,
    experience: int | None = None,
) -> dict:
    token_hash = _hash_token(token)
    invitation = TeacherRepository.get_active_invitation_by_token_hash(db, token_hash)

    if not invitation:
        raise ValueError("Invalid or already used invitation token.")

    if invitation.expires_at < datetime.utcnow():
        raise ValueError(
            "This invitation has expired. "
            "Please ask an admin to send a new one."
        )

    if UserRepository.get_by_email(db, invitation.email):
        raise ValueError("A user with this email already exists. Please contact admin.")

    if UserRepository.get_by_username(db, username):
        raise ValueError("Username already taken. Please choose another.")

    uid = generate_uid(db)

    user_data = {
        "uid": uid,
        "username": username,
        "email": invitation.email,
        "recovery_email": None,
        "password_hash": hash_password(password),
        "role": "teacher",
        "email_verified": True,
        "recovery_email_verified": False,
        "is_active": True,
        "deactivated_at": None
    }
    user = UserRepository.create(db, user_data)

    teacher_data = {
        "uid": uid,
        "name": name,
        "phone_number": phone_number,
        "specialization": specialization,
        "qualification": qualification,
        "experience": experience,
    }
    teacher = TeacherRepository.create(db, teacher_data)

    TeacherRepository.mark_invitation_used(db, invitation)

    welcome_email_sent = send_teacher_welcome_email(
        to_email=user.email,
        name=teacher.name,
        username=user.username
    )

    message = "Invitation accepted. Your teacher account has been created."
    if not welcome_email_sent:
        message += " However, the welcome email could not be sent."

    return {
        "message": message,
        "uid": user.uid,
        "username": user.username,
        "email": user.email,
        "role": user.role,
        "email_verified": user.email_verified,
        "teacher_id": teacher.teacher_id
    }

File: App(SpringBoot)/src/main/java/com/Project/LearningManagementSystem/service/InvitationService.java
package com.Project.LearningManagementSystem.service;

import com.Project.LearningManagementSystem.entity.Teacher;
import com.Project.LearningManagementSystem.entity.TeacherInvitation;
import com.Project.LearningManagementSystem.entity.User;
import com.Project.LearningManagementSystem.exception.BadRequestException;
import com.Project.LearningManagementSystem.repository.TeacherInvitationRepository;
import com.Project.LearningManagementSystem.repository.TeacherRepository;
import com.Project.LearningManagementSystem.repository.UserRepository;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class InvitationService {

    private final TeacherInvitationRepository invitationRepository;
    private final UserRepository userRepository;
    private final TeacherRepository teacherRepository;
    private final EmailService emailService;
    private final PasswordEncoder passwordEncoder;
    private final AuthService authService;

    private static final int INVITE_EXPIRY_HOURS = 48;

    @Transactional
    public Map<String, Object> createTeacherInvitation(String email, String invitedByUid, String acceptUrlBase) {
        email = email.trim().toLowerCase();

        if (userRepository.existsByEmail(email)) {
            throw new BadRequestException("A user with this email already exists.");
        }

        invitationRepository.findByEmailAndUsed(email, false).ifPresent(existing -> {
            if (existing.getExpiresAt().isAfter(LocalDateTime.now())) {
                throw new BadRequestException("An active invitation already exists for this email.");
            }
        });

        String token = UUID.randomUUID().toString();
        String tokenHash = AuthService.hashToken(token);

        TeacherInvitation invitation = new TeacherInvitation();
        invitation.setEmail(email);
        invitation.setTokenHash(tokenHash);
        invitation.setInvitedBy(invitedByUid);
        invitation.setUsed(false);
        invitation.setExpiresAt(LocalDateTime.now().plusHours(INVITE_EXPIRY_HOURS));
        invitation.setCreatedAt(LocalDateTime.now());
        invitationRepository.save(invitation);

        String baseUrl = (acceptUrlBase != null && !acceptUrlBase.isBlank()) ? acceptUrlBase : "http://localhost:3000/accept-invite";
        String inviteUrl = baseUrl + "?token=" + token;

        emailService.sendTeacherInvitationEmail(email, inviteUrl);

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Invitation sent to " + email);
        response.put("invitation_id", invitation.getInvitationId());
        response.put("email", email);
        response.put("expires_in_hours", INVITE_EXPIRY_HOURS);
        return response;
    }

    @Transactional
    public Map<String, Object> acceptTeacherInvitation(String token, String username, String password,
                                                      String name, String phoneNumber, String specialization,
                                                      String qualification, Integer experience) {
        String tokenHash = AuthService.hashToken(token);
        TeacherInvitation invitation = invitationRepository.findByTokenHash(tokenHash)
            .orElseThrow(() -> new BadRequestException("Invalid or already used invitation token."));

        if (invitation.isUsed()) {
            throw new BadRequestException("Invitation has already been used");
        }

        if (invitation.getExpiresAt().isBefore(LocalDateTime.now())) {
            throw new BadRequestException("This invitation has expired. Please ask an admin to send a new one.");
        }

        if (userRepository.existsByEmail(invitation.getEmail())) {
            throw new BadRequestException("A user with this email already exists.");
        }

        String uid = authService.generateUid();

        User user = new User();
        user.setUid(uid);
        user.setUsername(username.trim());
        user.setEmail(invitation.getEmail());
        user.setPasswordHash(passwordEncoder.encode(password));
        user.setRole("teacher");
        user.setEmailVerified(true);
        user.setActive(true);
        userRepository.save(user);

        Teacher teacher = new Teacher();
        teacher.setUid(uid);
        teacher.setName(name.trim());
        teacher.setPhoneNumber(phoneNumber);
        teacher.setSpecialization(specialization);
        teacher.setQualification(qualification);
        teacher.setExperience(experience);
        teacherRepository.save(teacher);

        invitation.setUsed(true);
        invitationRepository.save(invitation);

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Invitation accepted. Your teacher account has been created.");
        response.put("uid", user.getUid());
        response.put("username", user.getUsername());
        response.put("email", user.getEmail());
        response.put("role", user.getRole());
        response.put("email_verified", user.isEmailVerified());
        response.put("teacher_id", teacher.getTeacherId());
        return response;
    }
}
